File: BaekJoon/Python/그래프/p1197.py
import heapq
V, E = map(int, input().split())
adj = [[] for _ in range(V)]
chk = [False] * V
for i in range(E):
    a, b, c = map(int, input().split())
    adj[a-1].append((b, c))
    adj[b-1].append((a, c))

def prim():
    cnt = 0
    ans = 0
    PQ = []
    for i in range(len(adj[0])):
        heapq.heappush(PQ, (adj[0][i][1], 1, adj[0][i][0]))
    # print(heapq.heappop(PQ))
    chk[0] = True
    # prim(PQ)
    while True:
        cost, v1, v2 = heapq.heappop(PQ)
        if chk[v2-1]:
            continue
        chk[v2-1] = True
        cnt += 1
        ans += cost
        if cnt == V-1:
            break
        for i in range(len(adj[v2])):
            if not chk[adj[v2][i][0]]:
                heapq.heappush(PQ, (adj[v2][i][1], v2, adj[v2][i][0]))
    return ans
# print(chk)
print(prim())

File: BaekJoon/Java/그래프/p1197.java
package Algorithms;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class p1197 {

    static class Point {
        int u;
        int w;

        public Point(int u, int w) {
            this.u = u;
            this.w = w;
        }
    }

    static class Node implements Comparable<Node> {
        int v;
        int u;
        int w;

        public Node(int u, int v, int w) {
            this.v = v;
            this.u = u;
            this.w = w;
        }

        @Override
        public int compareTo(Node o) {
            return this.w - o.w;
        }
    }
    static int V;
    static int E;
    static ArrayList<ArrayList<Point>> adj = new ArrayList<>();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        V = Integer.parseInt(st.nextToken());
        E = Integer.parseInt(st.nextToken());

        for (int i = 0; i <= V; i++) {
            adj.add(new ArrayList<>());
        }

        for (int i = 0; i < E; i++) {
            st = new StringTokenizer(br.readLine());
            int A = Integer.parseInt(st.nextToken());
            int B = Integer.parseInt(st.nextToken());
            int C = Integer.parseInt(st.nextToken());
            adj.get(A).add(new Point(B, C));
            adj.get(B).add(new Point(A, C));
        }

        long weight = prim();

        System.out.println(weight);
    }

    private static long prim() {
        long weight = 0;
        PriorityQueue<Node> pq = new PriorityQueue<>();
        boolean[] chk = new boolean[V+1];
        for(Point p : adj.get(1)) {
            pq.offer(new Node(1, p.u, p.w));
        }
        chk[1] = true;
        int cnt = 0;
        while (true) {
            Node cur = pq.poll();
            int u = cur.u;
            int v = cur.v;
            if (chk[v]) continue;
            weight += cur.w;
            chk[v] = true;
            cnt++;
            if (cnt == V-1) break;
            for (Point p : adj.get(v)) {
                if (chk[p.u]) continue;
                pq.offer(new Node(v, p.u, p.w));
            }
        }
        return weight;
    }
}
